### PR TITLE
ADD: tool to generate last mnemonic word

### LINKED ISSUE #5680: New tool: generate a checksum word for a provided mnemonic seed

### PR DESCRIPTION
**Pull request by 3bitcoins**
Closes #5680 

Screenshot_1697197785

Screenshot_1697197935

### CODE REVIEW

**Comment by limpbrains** [comment]
Can you please add tests to the PR. Maybe extract checkMnemonic logic to a separate function and write a unit test for it.

Thank you for contributing to BlueWallet!

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @limpbrains] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/limpbrains

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by 3bitcoins** [comment]
@limpbrains I've moved the core logic to it's own file and added tests. Do you guys plan to ever review #5690 ? If so, I'll adapt it to use this helper instead of having this logic implemented twice.

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @limpbrains] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/limpbrains

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @limpbrains] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/limpbrains

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by limpbrains** [comment]
I've tried, works fine. 
I think UX can be better. 
- We need to show error if mnemonic is incorrect.
- show last word before button.
- Maybe text explanation of what is going on here whould be nice, user can be confused about why it is random. (I'm not sure)

**Comment by DrockDaniel** [comment]
I don't know how to do any of this it was a long time ago I created bitcoin

**Comment by 3bitcoins** [comment]
@limpbrains I've swapped the place of the result and the button, and added a message if the input is invalid. I also fixed a couple linter warnings i missed.

Based on how the other tools are, I don't think it's necessary to describe the exact usage of the tool in the screen itself. If you want, maybe some kind of documentation?

**Review by limpbrains — APPROVED** [review]

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by 3bitcoins** [comment]
@Overtorment Do you think you could review this?

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

File: tests/unit/checksumWords.test.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    }
+  });
+
+  it('generates 8 valid words for a 23 word input', () => {
+    const input =
+      'abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon ' +
+      'abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon';
+    const result = generateChecksumWords(input);
+    assert.ok(result);
+    assert.strictEqual(result.length, 8);
```
Comment (Overtorment) [inline tests/unit/checksumWords.test.js:23]: would be nice to check actual words generated

File: screen/wallets/generateWord.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    const possibleWords = generateChecksumWords(seedPhrase);
+
+    if (!possibleWords) {
+      // likely because of an invalid mnemonic
+      setResult(loc.autofill_word.error);
+      return;
+    }
+
+    const random = await randomBytes(1);
+    const randomindex = Math.round((random.readUInt8(0) / 255) * (possibleWords.length - 1));
```
Comment (Overtorment) [inline screen/wallets/generateWord.js:45]: `Math.random()` might have worked better here, without this awkward buffer-to-int conversion (`readUInt8(0) / 255`). not a deal breaker tho

File: blue_modules/checksumWords.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
+import * as bip39 from 'bip39';
+import createHash from 'create-hash';
+
+// partial (11 or 23 word) seed phrase
```
Comment (Overtorment) [inline blue_modules/checksumWords.js:4]: link to documentation would be nice

**Review by Overtorment — APPROVED** [review]
thanks for the contribution! merging this.
however,  theres room for improvement,  if you fancy doing a follow-up PR, most importantly, converting it to typescript (we now have a rule - every new file added might be in ts, while old js files are gradually converted)

File: screen/wallets/generateWord.js (hunk 2 above)
Comment (3bitcoins) [reply screen/wallets/generateWord.js:45]: I wasn't sure to what extent Math.random() was considered cryptographically secure in this project, and opted to use the safer RNG that i see used for other keygen purposes

File: tests/unit/checksumWords.test.js (hunk 1 above)
Comment (3bitcoins) [reply tests/unit/checksumWords.test.js:23]: logically it seems that if the right amount are generated and they're all valid, it's generated the right words

**Pull request merged**